File: benchmarks/network/TcpScatterGatherLoopback.c
#include <stdint.h>
#include <stddef.h>
#include <string.h>

#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>
#pragma comment(lib, "Ws2_32.lib")
typedef SOCKET socket_handle;
#else
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
typedef int socket_handle;
#endif

enum {
    FIRST_CHUNK_BYTES = 1536,
    SECOND_CHUNK_BYTES = 2560,
    CHUNK_BYTES = FIRST_CHUNK_BYTES + SECOND_CHUNK_BYTES,
    ITERATIONS = 256,
    EXPECTED_BYTES = CHUNK_BYTES * ITERATIONS
};

static uint16_t loopback_port(int attempt) {
#ifdef _WIN32
    long pid = (long)GetCurrentProcessId();
#else
    long pid = (long)getpid();
#endif
    if (pid < 0) {
        pid = 0;
    }

    return (uint16_t)(41000 + (pid % 20000) + attempt);
}

static int is_invalid_socket(socket_handle handle) {
#ifdef _WIN32
    return handle == INVALID_SOCKET;
#else
    return handle < 0;
#endif
}

static void close_socket(socket_handle handle) {
    if (is_invalid_socket(handle)) {
        return;
    }

#ifdef _WIN32
    closesocket(handle);
#else
    close(handle);
#endif
}

static socket_handle make_listener(uint16_t port) {
    socket_handle listener = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (is_invalid_socket(listener)) {
        return (socket_handle)-1;
    }

    struct sockaddr_in address;
    memset(&address, 0, sizeof(address));
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    address.sin_port = htons(port);

    if (bind(listener, (const struct sockaddr *)&address, sizeof(address)) != 0) {
        close_socket(listener);
        return (socket_handle)-1;
    }

    if (listen(listener, 128) != 0) {
        close_socket(listener);
        return (socket_handle)-1;
    }

    return listener;
}

static socket_handle connect_client(uint16_t port) {
    socket_handle client = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (is_invalid_socket(client)) {
        return (socket_handle)-1;
    }

    struct sockaddr_in address;
    memset(&address, 0, sizeof(address));
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    address.sin_port = htons(port);

    if (connect(client, (const struct sockaddr *)&address, sizeof(address)) != 0) {
        close_socket(client);
        return (socket_handle)-1;
    }

    return client;
}

static void consume(size_t *first_done, size_t first_len, size_t *second_done, size_t amount) {
    if (*first_done < first_len) {
        size_t remaining_first = first_len - *first_done;
        if (amount <= remaining_first) {
            *first_done += amount;
            return;
        }

        *first_done = first_len;
        amount -= remaining_first;
    }

    *second_done += amount;
}

static int write_all_vectored(socket_handle fd, const unsigned char *first, size_t first_len, const unsigned char *second, size_t second_len) {
    size_t first_written = 0;
    size_t second_written = 0;
    while (first_written < first_len || second_written < second_len) {
#ifdef _WIN32
        WSABUF buffers[2];
        DWORD buffer_count = 0;
        if (first_written < first_len) {
            buffers[buffer_count].buf = (CHAR *)(first + first_written);
            buffers[buffer_count].len = (ULONG)(first_len - first_written);
            buffer_count++;
        }

        if (second_written < second_len) {
            buffers[buffer_count].buf = (CHAR *)(second + second_written);
            buffers[buffer_count].len = (ULONG)(second_len - second_written);
            buffer_count++;
        }

        DWORD bytes_written = 0;
        if (WSASend(fd, buffers, buffer_count, &bytes_written, 0, NULL, NULL) != 0 || bytes_written == 0) {
            return 0;
        }

        consume(&first_written, first_len, &second_written, (size_t)bytes_written);
#else
        struct iovec buffers[2];
        int buffer_count = 0;
        if (first_written < first_len) {
            buffers[buffer_count].iov_base = (void *)(first + first_written);
            buffers[buffer_count].iov_len = first_len - first_written;
            buffer_count++;
        }

        if (second_written < second_len) {
            buffers[buffer_count].iov_base = (void *)(second + second_written);
            buffers[buffer_count].iov_len = second_len - second_written;
            buffer_count++;
        }

        ssize_t bytes_written = writev(fd, buffers, buffer_count);
        if (bytes_written <= 0) {
            return 0;
        }

        consume(&first_written, first_len, &second_written, (size_t)bytes_written);
#endif
    }

    return 1;
}

static int read_exact_vectored(socket_handle fd, unsigned char *first, size_t first_len, unsigned char *second, size_t second_len) {
    size_t first_read = 0;
    size_t second_read = 0;
    while (first_read < first_len || second_read < second_len) {
#ifdef _WIN32
        WSABUF buffers[2];
        DWORD buffer_count = 0;
        if (first_read < first_len) {
            buffers[buffer_count].buf = (CHAR *)(first + first_read);
            buffers[buffer_count].len = (ULONG)(first_len - first_read);
            buffer_count++;
        }

        if (second_read < second_len) {
            buffers[buffer_count].buf = (CHAR *)(second + second_read);
            buffers[buffer_count].len = (ULONG)(second_len - second_read);
            buffer_count++;
        }

        DWORD bytes_read = 0;
        DWORD flags = 0;
        if (WSARecv(fd, buffers, buffer_count, &bytes_read, &flags, NULL, NULL) != 0 || bytes_read == 0) {
            return 0;
        }

        consume(&first_read, first_len, &second_read, (size_t)bytes_read);
#else
        struct iovec buffers[2];
        int buffer_count = 0;
        if (first_read < first_len) {
            buffers[buffer_count].iov_base = first + first_read;
            buffers[buffer_count].iov_len = first_len - first_read;
            buffer_count++;
        }

        if (second_read < second_len) {
            buffers[buffer_count].iov_base = second + second_read;
            buffers[buffer_count].iov_len = second_len - second_read;
            buffer_count++;
        }

        ssize_t bytes_read = readv(fd, buffers, buffer_count);
        if (bytes_read <= 0) {
            return 0;
        }

        consume(&first_read, first_len, &second_read, (size_t)bytes_read);
#endif
    }

    return 1;
}

int main(void) {
#ifdef _WIN32
    WSADATA data;
    if (WSAStartup(MAKEWORD(2, 2), &data) != 0) {
        return 9;
    }
#endif

    unsigned char write_first[FIRST_CHUNK_BYTES];
    unsigned char write_second[SECOND_CHUNK_BYTES];
    unsigned char read_first[FIRST_CHUNK_BYTES];
    unsigned char read_second[SECOND_CHUNK_BYTES];
    memset(write_first, 0, sizeof(write_first));
    memset(write_second, 0, sizeof(write_second));
    memset(read_first, 0, sizeof(read_first));
    memset(read_second, 0, sizeof(read_second));

    socket_handle listener = (socket_handle)-1;
    uint16_t port = 0;
    for (int attempt = 0; attempt < 64; attempt++) {
        port = loopback_port(attempt);
        listener = make_listener(port);
        if (!is_invalid_socket(listener)) {
            break;
        }
    }

    if (is_invalid_socket(listener)) {
        return 1;
    }

    socket_handle client = connect_client(port);
    if (is_invalid_socket(client)) {
        close_socket(listener);
        return 2;
    }

    socket_handle server = accept(listener, NULL, NULL);
    close_socket(listener);
    if (is_invalid_socket(server)) {
        close_socket(client);
        return 3;
    }

    size_t total_read = 0;
    for (int iteration = 0; iteration < ITERATIONS; iteration++) {
        if (!write_all_vectored(client, write_first, FIRST_CHUNK_BYTES, write_second, SECOND_CHUNK_BYTES)) {
            close_socket(server);
            close_socket(client);
            return 4;
        }

        if (!read_exact_vectored(server, read_first, FIRST_CHUNK_BYTES, read_second, SECOND_CHUNK_BYTES)) {
            close_socket(server);
            close_socket(client);
            return 5;
        }

        total_read += CHUNK_BYTES;
    }

#ifdef _WIN32
    shutdown(client, SD_SEND);
#else
    shutdown(client, SHUT_WR);
#endif
    close_socket(client);
    close_socket(server);

#ifdef _WIN32
    WSACleanup();
#endif

    if (total_read != EXPECTED_BYTES) {
        return 6;
    }

    return 0;
}

File: benchmarks/network/TcpScatterGatherLoopback.rs
use std::io::{IoSlice, IoSliceMut, Read, Write};
use std::net::{Ipv4Addr, Shutdown, SocketAddrV4, TcpListener, TcpStream};
use std::process;

const FIRST_CHUNK_BYTES: usize = 1536;
const SECOND_CHUNK_BYTES: usize = 2560;
const CHUNK_BYTES: usize = FIRST_CHUNK_BYTES + SECOND_CHUNK_BYTES;
const ITERATIONS: usize = 256;
const EXPECTED_BYTES: usize = CHUNK_BYTES * ITERATIONS;

fn endpoint(attempt: u16) -> SocketAddrV4 {
    let pid = process::id();
    let port = 41_000 + (pid % 20_000) as u16 + attempt;
    SocketAddrV4::new(Ipv4Addr::LOCALHOST, port)
}

fn consume(first_done: &mut usize, first_len: usize, second_done: &mut usize, mut amount: usize) {
    if *first_done < first_len {
        let remaining_first = first_len - *first_done;
        if amount <= remaining_first {
            *first_done += amount;
            return;
        }

        *first_done = first_len;
        amount -= remaining_first;
    }

    *second_done += amount;
}

fn write_all_vectored(stream: &mut TcpStream, first: &[u8], second: &[u8]) -> Result<(), i32> {
    let mut first_written = 0_usize;
    let mut second_written = 0_usize;
    while first_written < first.len() || second_written < second.len() {
        let written = if first_written < first.len() {
            let slices = [
                IoSlice::new(&first[first_written..]),
                IoSlice::new(&second[second_written..]),
            ];
            stream.write_vectored(&slices).map_err(|_| 4)?
        } else {
            stream.write(&second[second_written..]).map_err(|_| 4)?
        };

        if written == 0 {
            return Err(4);
        }

        consume(
            &mut first_written,
            first.len(),
            &mut second_written,
            written,
        );
    }

    Ok(())
}

fn read_exact_vectored(stream: &mut TcpStream, first: &mut [u8], second: &mut [u8]) -> Result<(), i32> {
    let mut first_read = 0_usize;
    let mut second_read = 0_usize;
    while first_read < first.len() || second_read < second.len() {
        let read = if first_read < first.len() {
            let mut slices = [
                IoSliceMut::new(&mut first[first_read..]),
                IoSliceMut::new(&mut second[second_read..]),
            ];
            stream.read_vectored(&mut slices).map_err(|_| 5)?
        } else {
            stream.read(&mut second[second_read..]).map_err(|_| 5)?
        };

        if read == 0 {
            return Err(5);
        }

        consume(&mut first_read, first.len(), &mut second_read, read);
    }

    Ok(())
}

fn run() -> Result<(), i32> {
    let write_first = [0_u8; FIRST_CHUNK_BYTES];
    let write_second = [0_u8; SECOND_CHUNK_BYTES];
    let mut read_first = [0_u8; FIRST_CHUNK_BYTES];
    let mut read_second = [0_u8; SECOND_CHUNK_BYTES];
    let mut listener = None;
    let mut bound_endpoint = endpoint(0);
    for attempt in 0..64 {
        bound_endpoint = endpoint(attempt);
        match TcpListener::bind(bound_endpoint) {
            Ok(value) => {
                listener = Some(value);
                break;
            }
            Err(_) => {}
        }
    }

    let listener = listener.ok_or(1)?;
    let mut client = TcpStream::connect(bound_endpoint).map_err(|_| 2)?;
    let (mut server, _) = listener.accept().map_err(|_| 3)?;
    drop(listener);

    let mut total_read = 0_usize;
    for _ in 0..ITERATIONS {
        write_all_vectored(&mut client, &write_first, &write_second)?;
        read_exact_vectored(&mut server, &mut read_first, &mut read_second)?;
        total_read += CHUNK_BYTES;
    }

    client.shutdown(Shutdown::Write).map_err(|_| 6)?;
    if total_read != EXPECTED_BYTES {
        return Err(7);
    }

    Ok(())
}

fn main() {
    if let Err(code) = run() {
        process::exit(code);
    }
}
